### ISSUE TITLE
When a link that leads to another page inside the same container is opened, which then redirects to another page assigned to a different container, sometimes the page is opened twice.

### ISSUE THREAD

**Issue by magkopian** [mozilla/multi-account-containers#940]
I apologize for the long title, but I didn’t know how exactly to describe this using fewer words. I did a [screen capture](https://www.youtube.com/watch?v=Qu4CMSuR0Zw) which I hope helps demonstrating what exactly happens.

In this instance, I have `www.youtube.com` assigned to always open inside the `Google` container, while `www.thingiverse.com` is assigned on the `Personal` container.

YouTube replaces the links inside the descriptions of videos with its own, which then redirect to the actual destination. For example, when I click on the link in the video, this opens the following URL on a new tab,

```
https://www.youtube.com/redirect?event=video_description&v=NiY8umNK0X8&q=http%3A%2F%2Fwww.thingiverse.com%2Fthing%3A952987&redir_token=PvLq3E6oUcBOJAGxdcXCMIbMIJ58MTUxMDMzNjc1N0AxNTEwMjUwMzU3
```
which then redirects to,

```
http://www.thingiverse.com/thing:952987
```

The weird thing, is that the issue appears to happen at random. As you can also see in the video, sometimes the page will open twice while others it won't.

### Application Basics

Attribute | Value
|-----------|--------|
Name | Firefox
Version | 58.0b1
Build ID | 20171103003834
Update Channel | aurora
User Agent | Mozilla/5.0 (X11; Linux x86_64; rv:58.0) Gecko/20100101 Firefox/58.0
OS | Linux 4.9.0-4-amd64
Multiprocess Windows | 2/2 (Enabled by default)
Web Content Processes | 4/4
Stylo | content = true (enabled by default), chrome = false (disabled by default)
Google Key | Found
Mozilla Location Service Key | Found
Safe Mode | false

### Developer Edition Features

Name    | Version 	| ID
|----------|-------------|------|
Activity Stream	| 2017.10.20.1325-95b4c353	| activity-stream@mozilla.org
Application Update Service Helper | 2.0 | aushelper@mozilla.org
Firefox Screenshots	 | 19.2.0 | screenshots@mozilla.org
Follow-on Search Telemetry | 0.9.6 | followonsearch@mozilla.com
Form Autofill | 1.0 | formautofill@mozilla.org
Photon onboarding | 1.0 | onboarding@mozilla.org
Pocket | 1.0.5 | firefox@getpocket.com
Shield Recipe Client | 76 | shield-recipe-client@mozilla.org
Web Compat | 1.1 | webcompat@mozilla.org
WebCompat Reporter | 1.0.0 | webcompat-reporter@mozilla.org

### Extensions

Name  | Version  | Enabled  | ID
|-----|--------|---------|---|
ADB Helper | 0.11.2 | true | adbhelper@mozilla.org
Command Line Get (CLG) | 0.0.4 | true | mycliget@example.com
Google search link fix | 1.6.5 | true | jid0-XWJxt5VvCXkKzQK99PhZqAn7Xbg@jetpack
HTTPS Everywhere | 2017.10.30 | true | https-everywhere@eff.org
I don't care about cookies | 2.7.3 | true | jid1-KKzOGWgsW3Ao4Q@jetpack
KeePassHttp-Connector | 1.0.3 | true | keepasshttp-connector@addons.brandt.tech
Multi-Account Containers | 4.0.3 | true | @testpilot-containers
Open Tabs Next to Current | 2.0.8 | true | opentabsnexttocurrent@sblask
uBlock Origin | 1.14.18 | true | uBlock0@raymondhill.net
Vue.js devtools | 3.1.6 | true | {5caff8cc-3d2e-4110-a88a-003cc85b3858}
Spell checker based on Voikko | 2.2 | false | {b676e3ff-cda7-4e0c-b2b8-74e4bb40a67a}

**Comment by gaborluk** [comment]
I can consistently replicate this issue, and this would be solved by #1029 as well.

**Comment by stoically** [comment]
What I noticed while working on WebExtensions is, that other Add-ons that do redirects - such as HTTPS Everywhere - effectively *sometimes* are faster in their `onBeforeRequest` handler and thus another request is issued that is then again seen by the Add-on. At first I thought it's a bug, but a Firefox dev stated:

```When multiple blocking handlers modify a request, only one set of modifications take effect. Redirects and cancellations have the same precedence.```

I fixed this by saving the `requestId` that gets canceled and if another request with that `requestId` comes in, I cancel that too. That's because redirected requests have the same requestId. This approach requires to set a cleanup timeout for the saved canceled requestIds though.

I'll look into providing a PR.
